Library/Homebrew/cmd/install.rb

```diff
@@ -135,13 +135,41 @@ def install
           raise "No devel block is defined for #{f.full_name}"
         end
 
-        current = f if f.installed?
-        current ||= f.old_installed_formulae.first
+        if f.keg_only? && !f.installed_prefixes.empty? && f.opt_prefix.symlink? && !ARGV.force?
+          # keg-only install is only possible when no ohter version is
+          # optlinked, because silent install can break dependencies. Therefore
+          # before performing other checks we need to be sure --force flag is
+          # passed.
+          opoo "#{f.full_name} is a keg-only and another version is optlinked."
+          puts "Use `brew install --force` if you want to install this version"
+        elsif (ARGV.build_head? && (version = f.latest_head_version) &&
+            !f.head_version_outdated?(version, fetch_head: ARGV.fetch_head?)) ||
+              (f.prefix.exist? && !f.prefix.children.empty?)
+          # After we're sure that --force flag is passed for optlinked keg-only
+          # we need to be sure that the version we're attempting to install
+          # is not already installed.
 
-        if current
-          msg = "#{current.full_name}-#{current.installed_version} already installed"
-          unless current.linked_keg.symlink? || current.keg_only?
-            msg << ", it's just not linked"
+          installed_version = if ARGV.build_head?
+            f.latest_head_version
+          else
+            f.pkg_version
+          end
+
+          msg = "#{f.full_name}-#{installed_version} already installed"
+          if f.linked_keg.symlink? && f.linked_keg.resolved_path != f.prefix(installed_version)
```

Comment:
`f.linked_keg.resolved_path != f.prefix(installed_version)` could be a helper method or just variable.